docs/DesignDocs/SGX_EMM.md

```diff
@@ -0,0 +1,350 @@
+
+The clients can also call the EMM commit API, sgx_mm_commit, to proactively commit specific sub-regions in a COMMIT_ON_DEMAND allocation to avoid future page fault.
+
+Some EMM clients, <i>e.g.</i>, a dynamic code loader wishing to load code on page fault, can register a custom handler for page fault at the time of allocation request. In the custom page fault handler, it can invoke an API, sgx_mm_commit_data, to commit and load data to newly committed EPC page at the same time as supported by EACCEPTCOPY. 
+
+Committed pages will stay committed (regardless how they were committed) until the clients calls the uncommit API, sgx_mm_uncommit, on them or the allocation they belong to is deallocated by sgx_mm_dealloc.    
+
+**Modify Page Attributes**
+
+The EMM clients may call sgx_mm_modify/sgx_mm_modify_ex to request permissions and/or page type changes for pages in existing allocations.
+
+
+## Notes on Internal Design ##
+
+The enclave memory manager keeps track of memory allocation and layout info inside enclave address range (ELRANGE) using an internal structure called Enclave Memory Area (EMA) List. The EMA and EMA list are considered private data structure of memory manager, and their internals are not exposed in client-facing APIs.
+- The EMA list tracks all memory regions in use (reserved, committed, commit-on-demand) in ELRANGE.
+- Ranges in ELRANGE not tracked by an EMA are considered free and ready for new allocations
+- The EMM labels certain EMAs reserved for runtime or its internal usage and make them not accessible from public APIs.
+- A thread calling an EMM API on an EMA with operation pending in another thread will wait until the pending operation is finished. 
+
+**Assumptions:**
+
+- When enclave is loaded, the OS has reserved the whole address range covered by ELRANGE. It is assumed the host app will not remap any part of this reserved range.
+  - In future, when the new feature that allows loading enclave with base address at zero, we may support partial ELRANGE reserved by the OS.  
+- Memory manager does not check EPC pressure, or proactively trim pages when EPC runs low. The OS can reclaim EPC pages when EPC running low or cgroups threshold reached
+- Memory manager does not maintain and recycle committed then freed pages
+  - Whenever a page is freed (via dealloc or uncommit API), it is trimmed from enclave and need be re-allocated and committed before re-use.
+  - Owner of a region can re-purpose a sub-region of it by calling sgx_mm_modify_type/permissions to split out the sub-region to be reused.
+- Memory manager does not call back client for #GP handling. Memory manager code will ensure itself would not cause #GP, and only register a #PF handler with the enclave global exception handler registry. Clients wish to handle #GP can register its own exception handler with the global handler registry.
+
+
+Public APIs
+-----------------
+
+### sgx_mm_alloc
```

Comment:
Also, are these proposed as APIs to be added to OE?  (where APIs would be prefixed with `oe_`)

Reply:
The intent is to reuse implementation of these APIs between OE and Intel SGX sdk. I'm not very familiar with OE yet. From what I see, if we decide to expose these as common APIs across different technologies for OE, then we need add these APIs with oe_ prefix, and do weak alias for SGX impl?

Reply:
Do you really need a weak alias?  I suspect a simple #define macro that defines one to be an alias for the other, would be sufficient.

Reply:
Also I don't see any text in this document about "reuse implementation".   Please explain the code relationship.  Does one depend on the other or should it be in a separate common library consumed by both?

Reply:
Per discussion in arch forum, we plan to reuse it as lib